### ISSUE TITLE
Set lhs of distance to cluster centers in Hierarchical clustering in store_model_features

### ISSUE THREAD

**Issue by vinx13** [shogun-toolbox/shogun#4388]
every [`CDistanceMachine`](https://github.com/shogun-toolbox/shogun/blob/58b914a6b326cd5f799727e39737bc4b4b11b286/src/shogun/machine/DistanceMachine.h#L105) has to make sure that cluster centers are in lhs of distance variable, the task is to fix the todo in https://github.com/shogun-toolbox/shogun/blob/58b914a6b326cd5f799727e39737bc4b4b11b286/src/shogun/clustering/Hierarchical.h#L122

see [KMeans](https://github.com/shogun-toolbox/shogun/blob/58b914a6b326cd5f799727e39737bc4b4b11b286/src/shogun/clustering/KMeansBase.cpp#L249) for example

**Comment by karlnapf** [comment]
as if you need details

**Comment by braceletboy** [comment]
Hi @karlnapf ,
I was looking at the Hierarchical.cpp file while trying to look into this issue. I think there is something wrong with this [line](https://github.com/shogun-toolbox/shogun/blob/58b914a6b326cd5f799727e39737bc4b4b11b286/src/shogun/clustering/Hierarchical.cpp#L129).  
I think the condition should be `for(; l<num && l<=merges && k<num_pairs-1; l++)`. I tried it out on a small example manually and checked myself. Correct me if I am wrong. Also, if the condition `num-l>=merges` is true then so is `l<num`.

**Comment by karlnapf** [comment]
This seems very reasonable to me.
Would you like to send a PR with an update.
In addition, we should maybe test some results against another library on a very simple example?

**Comment by karlnapf** [comment]
@braceletboy this is unrelated to this issue though, so pls let's communicate in another issue, or even better, in a PR from your side

**Comment by braceletboy** [comment]
@karlnapf Sure. I am on it.

**Comment by braceletboy** [comment]
@karlnapf , I was trying to use the `set_feature_vector()` function in the `CDenseFeatures` class. Which is documented [here](http://www.shogun-toolbox.org/api/latest/classshogun_1_1CDenseFeatures.html#affd56f6e718d53ead36a9b67e53055cc)

But, I couldn't find it in the source code. Refer [DenseFeatures.h](https://github.com/shogun-toolbox/shogun/blob/58b914a6b326cd5f799727e39737bc4b4b11b286/src/shogun/features/DenseFeatures.h). Has it been removed?

**Comment by gf712** [comment]
@braceletboy it seems like it was removed [here](
https://github.com/shogun-toolbox/shogun/commit/f10f3ef2b28b7a5f0e137b3e4a7f8abf80a64201#diff-0a19a445c0877308de5ae341f137cc17).

**Comment by vinx13** [comment]
@braceletboy Yes we removed it because we are making features immutable.

**Comment by karlnapf** [comment]
Yep, just confirming that we don't want to allow modification of features. You can of course get a pointer to the matrix and modify memory directly, but we don't want to encourage that via the API

**Comment by braceletboy** [comment]
@karlnapf  Yeah. Finally I had to use a pointer. Thanks

**Comment by braceletboy** [comment]
@vinx13 Are we also supposed to add the get_cluster_centers and get_radiuses functions? I see that these functions were implemented in the [CKmeans](http://shogun.ml/api/latest/classshogun_1_1CKMeansBase.html#a5d8a09aeadada018747786a5470d3653) class and that they were used in the meta examples

**Comment by vinx13** [comment]
@braceletboy They should be registered using tag framework https://github.com/shogun-toolbox/shogun/blob/58b914a6b326cd5f799727e39737bc4b4b11b286/src/shogun/clustering/KMeansBase.cpp#L365 and then we can get them using `get`. The getter functions should be removed in the future.

**Comment by braceletboy** [comment]
Umm okay. I don't know if this is a good question. If they are to be removed in the future, why are they being written now? Also I used the get_feature_vector function of CDenseFeatures will this cause a problem in the future?

**Comment by braceletboy** [comment]
Okay. I will have a look at it and change my code. Thanks

**Comment by karlnapf** [comment]
Note that in KMeansBase, the function does something rather than just returning the cluster centres. In this case it is rather pointless as it just extracts the feature matrix of the lhs features.

What we eventually want is a way to extract/compute the cluster centres from the `put/get` API. The classes KMeans and Hierarchical will eventually be hidden from the interfaces so no methods will be available.

**Comment by braceletboy** [comment]
@karlnapf In the `get_cluster_centers` of [KMeansBase](https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/clustering/KMeansBase.cpp#L222) also, we are extracting the feature matrix of lhs features.

**Comment by braceletboy** [comment]
#4510

**Comment by braceletboy** [comment]
@karlnapf  @vinx13  Referencing the pull request I sent here: #4510
